### PR TITLE
json: Support custom quantities

### PR DESCRIPTION
**Pull request by giantmustache**
The Activator is used to instantiate objects, instead of calling Quantity.From.
The assumption is that there will be a constructor of new T(double value, TUnit unit).

More info in issue #934

### CODE REVIEW

**Comment by angularsen** [comment]
Thanks, I'll try to get around to this in a couple of days.

File: UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs [hunk 1]
```diff
@@ -87,7 +116,40 @@ protected IQuantity ConvertValueUnit(ValueUnit valueUnit)
...
+            (var unitEnumTypeName, var unitEnumValue) = SplitUnitString(unit);
+            return _registeredTypes.Find(t => t.Unit.Name == unitEnumTypeName).Quantity;
+        }
+
+        private Enum GetUnit(string unit)
+        {
+            (var unitEnumTypeName, var unitEnumValue) = SplitUnitString(unit);
+
+            // First try to find the name in the list of registered types.
+            var unitEnumType = _registeredTypes.Find(t => t.Unit.Name == unitEnumTypeName).Unit;
```
Comment (angularsen) [inline UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs:130]: It seems like a small weakness in the current design that the namespace is not included in the serialization. But I guess for now, this is what we have to work with.

**Review by angularsen — CHANGES_REQUESTED** [review]
Great work! Just some minor things.

File: UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs [hunk 2]
```diff
@@ -17,6 +18,28 @@ namespace UnitsNet.Serialization.JsonNet
...
+        /// <summary>
+        /// Register custom types so that the converter can instantiate these quantities.
+        /// Instead of calling <see cref="Quantity.From"/>, the <see cref="Activator"/> will be used to instantiate the object.
+        /// It is therefore assumed that the constructor of <paramref name="quantity"/> is specified with <c>new T(double value, typeof(<paramref name="unit"/>) unit)</c>.
+        /// </summary>
+        public void RegisterCustomType(Type quantity, Type unit)
+        {
+            if (!typeof(T).IsAssignableFrom(quantity))
+            {
+                throw new ArgumentException($"The type {quantity} is not a {nameof(T)}");
```
[the remark this answers is not in the document]
Comment (giantmustache) [reply UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs:32]: Fixed

File: UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs [hunk 3]
```diff
@@ -87,7 +116,40 @@ protected IQuantity ConvertValueUnit(ValueUnit valueUnit)
...
+        }
+
+        private Enum GetUnit(string unit)
+        {
+            (var unitEnumTypeName, var unitEnumValue) = SplitUnitString(unit);
+
+            // First try to find the name in the list of registered types.
+            var unitEnumType = _registeredTypes.Find(t => t.Unit.Name == unitEnumTypeName).Unit;
+
+            if (unitEnumType is null)
```
[the remark this answers is not in the document]
Comment (giantmustache) [reply UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs:132]: Do you mean that you would first like to search in `UnitsNet.Units.MassUnit,UnitsNet` and if you get an exception, search in the custom types?

File: UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs [hunk 4]
```diff
@@ -17,6 +18,28 @@ namespace UnitsNet.Serialization.JsonNet
...
+            {
+                throw new ArgumentException($"The type {quantity} is not a {nameof(T)}");
+            }
+
+            if (!typeof(Enum).IsAssignableFrom(unit))
+            {
+                throw new ArgumentException($"The type {unit} is not a {nameof(Enum)}");
+            }
+
+            _registeredTypes.Add((quantity, unit));
```
[the remark this answers is not in the document]
Comment (giantmustache) [reply UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs:40]: Done, using the unit as key in the dictionary.

File: UnitsNet.Tests/CustomQuantities/HowMuchTests.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        [Fact]
+        public static void SerializeAndDeserializeCreatesSameObjectForIQuantity()
+        {
+            var jsonSerializerSettings = new JsonSerializerSettings { Formatting = Formatting.Indented };
+            var quantityConverter = new UnitsNetIQuantityJsonConverter();
+            quantityConverter.RegisterCustomType(typeof(HowMuch), typeof(HowMuchUnit));
+            jsonSerializerSettings.Converters.Add(quantityConverter);
+
+            var quantity = new HowMuch(12.34, HowMuchUnit.ATon);
```
[the remark this answers is not in the document]
Comment (giantmustache) [reply UnitsNet.Tests/CustomQuantities/HowMuchTests.cs:17]: I moved it, but `HowMuch` is situated in `UnitsNet.Tests`, so I had to add a reference `UnitsNet.Tests` to `UnitsNet.Serialization.JsonNet.Tests`

File: UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (angularsen) [reply UnitsNet.Serialization.JsonNet/UnitsNetBaseJsonConverter.cs:138]: Looking at the code now, I don't recognize what I meant 😅

**Comment by angularsen** [comment]
[Release JsonNet/4.4.0 · angularsen/UnitsNet](https://github.com/angularsen/UnitsNet/releases/tag/JsonNet%2F4.4.0)

**Comment by angularsen** [comment]
@giantmustache nuget 4.4.0 is on the way out, it would be great if you could give it a try and report back that it works.

Queued build:
https://ci.appveyor.com/project/angularsen/unitsnet/builds/40017368

**Pull request merged**